File: frontend/src/types/matterCreation.ts
/**
 * Types for Matter Creation (Intake) - POST /api/matters
 */

export interface MatterParticipantCreate {
  participant_name: string;
  email_address?: string | null;
  organization?: string | null;
  role_relationship?: string | null;
  is_active: boolean;
}

export interface IntakeNarrative {
  update_summary: string;
  source_actor?: string | null;
  source_reference?: string | null;
  occurred_at?: string | null;  // ISO datetime string
  logged_by?: string | null;
}

export interface MatterCreateRequest {
  matter_key: string;
  client_id: string;
  matter_id: string;
  client_name: string;
  matter_name: string;
  matter_description: string;
  matter_status: 'open' | 'closed' | 'pending' | 'suspended';
  practice_area?: string | null;
  matter_type?: string | null;
  matter_aliases_identifiers?: string | null;
  primary_attorney?: string | null;
  participants?: MatterParticipantCreate[];
  intake_narrative?: IntakeNarrative | null;
}

export interface MatterParticipantSummary {
  participant_id: number;
  matter_key: string;
  participant_name: string;
  email_address: string | null;
  organization: string | null;
  role_relationship: string | null;
  is_active: boolean;
}

export interface MatterCaseBrainSummary {
  brain_entry_id: number;
  email_id: string | null;
  occurred_at: string;
  source_type: string;
  source_reference: string | null;
  source_actor: string | null;
  update_summary: string;
  logged_by: string | null;
}

export interface MatterCreateResponse {
  matter_key: string;
  client_id: string;
  matter_id: string;
  client_name: string;
  matter_name: string;
  practice_area: string | null;
  matter_type: string | null;
  matter_description: string;
  matter_aliases_identifiers: string | null;
  matter_status: string;
  primary_attorney: string | null;
  last_brain_update: string | null;
  created_at: string;
  updated_at: string;
  participants: MatterParticipantSummary[];
  case_brain_entries: MatterCaseBrainSummary[];
}

File: backend/app/schemas/matter_creation.py
"""
Pydantic schemas for the Matter Creation (Intake) endpoint.

POST /api/matters
"""

from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field, field_validator


# ---------------------------------------------------------------------------
# Participant
# ---------------------------------------------------------------------------
class MatterParticipantCreate(BaseModel):
    """Initial participant for a newly created Matter."""

    participant_name: str = Field(..., min_length=1)
    email_address: Optional[str] = None
    organization: Optional[str] = None
    role_relationship: Optional[str] = None
    is_active: bool = True


# ---------------------------------------------------------------------------
# Intake narrative (optional opening Case Brain entry)
# ---------------------------------------------------------------------------
class IntakeNarrative(BaseModel):
    """Optional intake narrative written as the first Case Brain entry."""

    update_summary: str = Field(..., min_length=1)
    source_actor: Optional[str] = None
    source_reference: Optional[str] = None
    occurred_at: Optional[datetime] = None
    logged_by: Optional[str] = None

    @field_validator("update_summary")
    @classmethod
    def _reject_blank(cls, value: str) -> str:
        if not value or not value.strip():
            raise ValueError("update_summary must not be empty or whitespace-only.")
        return value


# ---------------------------------------------------------------------------
# Request
# ---------------------------------------------------------------------------
class MatterCreateRequest(BaseModel):
    """Request body for POST /api/matters."""

    matter_key: str = Field(..., min_length=1, max_length=50)
    client_id: str = Field(..., min_length=1, max_length=30)
    matter_id: str = Field(..., min_length=1, max_length=30)
    client_name: str = Field(..., min_length=1, max_length=200)
    matter_name: str = Field(..., min_length=1, max_length=200)
    matter_description: str = Field(..., min_length=1, max_length=500)
    matter_status: str = Field(default="open", pattern=r"^(open|closed|pending|suspended)$")
    practice_area: Optional[str] = Field(default=None, max_length=100)
    matter_type: Optional[str] = Field(default=None, max_length=100)
    matter_aliases_identifiers: Optional[str] = None
    primary_attorney: Optional[str] = Field(default=None, max_length=200)
    participants: Optional[List[MatterParticipantCreate]] = None
    intake_narrative: Optional[IntakeNarrative] = None


# ---------------------------------------------------------------------------
# Response (mirrors MatterResponse + nested collections)
# ---------------------------------------------------------------------------
class MatterParticipantSummary(BaseModel):
    participant_id: int
    matter_key: str
    participant_name: str
    email_address: Optional[str]
    organization: Optional[str]
    role_relationship: Optional[str]
    is_active: bool


class MatterCaseBrainSummary(BaseModel):
    brain_entry_id: int
    email_id: Optional[str]
    occurred_at: datetime
    source_type: str
    source_reference: Optional[str]
    source_actor: Optional[str]
    update_summary: str
    logged_by: Optional[str]


class MatterCreateResponse(BaseModel):
    matter_key: str
    client_id: str
    matter_id: str
    client_name: str
    matter_name: str
    practice_area: Optional[str]
    matter_type: Optional[str]
    matter_description: str
    matter_aliases_identifiers: Optional[str]
    matter_status: str
    primary_attorney: Optional[str]
    last_brain_update: Optional[datetime]
    created_at: datetime
    updated_at: datetime
    participants: List[MatterParticipantSummary] = []
    case_brain_entries: List[MatterCaseBrainSummary] = []
